### ISSUE TITLE
cython extensions are compiled to c++ source code.

### ISSUE THREAD

**Issue by rainwoodman** [astropy/halotools#828]
It appears that halotools doesn't really need c++. If that's the case, using C makes life of downstream packaging a lot easier.

**Comment by aphearin** [comment]
I'd love to get rid of the c++ compilation @rainwoodman - The only place where any cpp-specific feature is used is in the pairwise distance engines: `pairwise_distance_3d_engine`, `pairwise_distance_xy_z_engine` and `conditional_pairwise_distances`. In particular, these functions use `libcpp.vector`. I have not looked into how much the algorithm would need to change if this dependence were dropped - @duncandc could you comment to that effect?

**Comment by duncandc** [comment]
I agree that it would be ideal to get rid of the C++ dependency.  Let me take a look at those functions to see if we can get around using the `libcpp.vector` library without any significant penalty.

**Comment by rainwoodman** [comment]
Sorry I didn't realize it is using libcpp.vector.

**Comment by aphearin** [comment]
Those three functions are the only ones using libcpp.vector in the entire repo @rainwoodman - if you know of common workarounds for this, I would happily use those instead. These three functions do not need to be ultra-high performance.

**Comment by rainwoodman** [comment]
If there is a reasonably sized upper bound, can you simply allocate according to the bound and update a `size` variable?

incremental growth of around 1.5x with `realloc()` is not very hard too, you only need to keep size and maxsize.
